nixui/options/parser.py

```diff
@@ -1,40 +1,95 @@
+                syntax_tree.Token(id=uuid.uuid4(), name='INJECTION', position=None, quoted=comment_str),
+                index=node_to_prefix_comment.elems.index(tree.get_parent(token, node=True))
+            )
+        # add new option definitions where they don't exist
+        else:
+            attr_str = '.'.join(attr_loc)
+            quoted = f'{attr_str} = {option_expr_map[attr_loc]};'
+            token = syntax_tree.Token(id=uuid.uuid4(), name='INJECTION', position=None, quoted=quoted)
+            tree.insert(returned_attr_set_node, token, index=1)
+            # insert comment
+            tree.insert(
+                tree.get_parent(token, node=True),
+                syntax_tree.Token(id=uuid.uuid4(), name='INJECTION', position=None, quoted=comment_str),
+                index=1
+            )
+    return tree.to_string()
+
+
+def apply_indentation(string, num_spaces):
+    return '\n'.join([
+        (' ' * num_spaces) + line
+        for line in string.split('\n')
+    ])
+
 
 def get_all_option_values(root_module_path):
     option_expr_map = {}
     for module_path in [root_module_path]:
-        get_col_line_index = get_column_line_index_map(module_path)
-        tree = syntax_tree.get_syntax_tree(module_path)
-        for attr_loc, attr_data in nix_eval.get_modules_defined_attrs(module_path).items():
-            character_index = get_col_line_index(
-                attr_data['position']['line'] - 1,
-                attr_data['position']['column'] - 1
-            )  # 'line' and 'column' are 1-indexed
-            definition_node = syntax_tree.get_node_at_position(
-                tree,
-                character_index,
-                'NODE_KEY_VALUE'
-            )
-            key_node, value_node = [e for e in definition_node.elems if isinstance(e, syntax_tree.Node)]
-            value_expr = syntax_tree.get_full_node_string(value_node)
+        tree = syntax_tree.SyntaxTree(module_path)
+        for attr_loc, (key_node, value_node) in get_key_value_nodes(module_path, tree).items():
+            value_expr = tree.to_string(value_node)
             option_expr_map[attr_loc] = value_expr
     return option_expr_map
 
 
-def get_column_line_index_map(path):
-    line_index_map = {}
-    index = 0
-    with open(path) as f:
-        for i, line in enumerate(f.readlines()):
-            line_index_map[i] = index
-            index += len(line.encode("utf8"))
+def get_imported_modules(module_path):
+    # TODO: fix this, possibly using rnix-lsp
+    return nix_eval.eval_attribute(module_path, "imports")
+
 
-    # No need to use a special column mapping, as Nix columns are already byte offsets
-    mapper = lambda line, col: line_index_map[line] + col
-    return mapper
+def get_returned_attr_set_node(module_path, tree):
+    """
+    Get the NODE_ATTR_SET containing the attributes which are returned by the module
+    """
+    # TODO: fix HACK, currently we assume the node containing `imports` is the returned attr set
```

Comment:
Not urgent, but this is not the case for canonical-form modules. Instead, we should check for the existence of the `config` attribute first.

Reply:
Added to this issue https://github.com/nix-gui/nix-gui/issues/48